rust/lance/src/dataset/fragment/write.rs

```diff
@@ -0,0 +1,278 @@
+use arrow_schema::Schema as ArrowSchema;
+use datafusion::execution::SendableRecordBatchStream;
+use futures::StreamExt;
+use lance_core::datatypes::Schema;
+use lance_core::Error;
+use lance_datafusion::chunker::chunk_stream;
+use lance_datafusion::utils::{peek_reader_schema, reader_to_stream};
+use lance_file::writer::FileWriter;
+use lance_io::object_store::ObjectStore;
+use lance_table::format::Fragment;
+use lance_table::io::manifest::ManifestDescribing;
+use snafu::{location, Location};
+use uuid::Uuid;
+
+use crate::dataset::builder::DatasetBuilder;
+use crate::dataset::{WriteMode, WriteParams, DATA_DIR};
+use crate::Result;
+
+/// Builder for writing a new fragment.
+///
+/// This builder can be re-used to write multiple fragments.
+pub struct FragmentCreateBuilder<'a> {
+    dataset_uri: &'a str,
+    schema: Option<&'a Schema>,
+    write_params: Option<&'a WriteParams>,
+}
+
+impl<'a> FragmentCreateBuilder<'a> {
+    pub fn new(dataset_uri: &'a str) -> Self {
+        Self {
+            dataset_uri,
+            schema: None,
+            write_params: None,
+        }
+    }
+
+    /// Set the schema of the fragment. If it is not known, it will be inferred.
+    ///
+    /// If the schema isn't provided, but the `write_mode` is `WriteMode::Append`,
+    /// the schema will be inferred from the existing dataset.
+    ///
+    /// If that fails, the schema will be inferred from the first batch.
+    pub fn schema(mut self, schema: &'a Schema) -> Self {
+        self.schema = Some(schema);
+        self
+    }
+
+    /// Set the write parameters.
+    pub fn write_params(mut self, params: &'a WriteParams) -> Self {
+        self.write_params = Some(params);
+        self
+    }
+
+    /// Write a fragment.
+    pub async fn write(
+        &self,
+        reader: impl RecordBatchReader + Send + 'static,
+        id: Option<u64>,
+    ) -> Result<Fragment> {
+        let (stream, schema) = self.get_stream_and_schema(Box::new(reader)).await?;
+        self.write_impl(stream, schema, id).await
+    }
+
+    async fn write_impl(
+        &self,
+        stream: SendableRecordBatchStream,
+        schema: Schema,
+        id: Option<u64>,
+    ) -> Result<Fragment> {
+        let id = id.unwrap_or_default();
+
+        let params = self.write_params.map(Cow::Borrowed).unwrap_or_default();
+        let progress = params.progress.as_ref();
+
+        Self::validate_schema(&schema, stream.schema().as_ref())?;
+
+        let (object_store, base_path) = ObjectStore::from_uri(self.dataset_uri).await?;
+        let filename = format!("{}.lance", Uuid::new_v4());
+        let mut fragment = Fragment::with_file_legacy(id, &filename, &schema, None);
+        let full_path = base_path.child(DATA_DIR).child(filename.clone());
+        let mut writer = FileWriter::<ManifestDescribing>::try_new(
+            &object_store,
+            &full_path,
+            schema,
+            &Default::default(),
+        )
+        .await?;
```

Comment:
This is no different than before but I suppose we will need to update this path to use the v2 writer as well.

Reply:
Good point. I've added a TODO to https://github.com/lancedb/lance/issues/1929